subprojects/core-api/src/main/java/org/gradle/api/provider/CollectionPropertyConfigurer.java

```diff
@@ -70,4 +69,83 @@ public interface CollectionPropertyConfigurer<T> extends ConfigurableValue.Confi
      * @param provider Provider of elements
      */
     void addAll(Provider<? extends Iterable<? extends T>> provider);
+
+    /**
+     * Adds an element to the property value.
+     *
+     * <p>
+     * Contrary to {@link #add(Object)}, this method operates on the
+     * actual value of the property, be it defined via a convention or explicit value.
+     * </p>
+     *
+     * @param element The element
+     * @since 8.7
+     */
+    @Incubating
+    void append(T element);
+
+    /**
+     * Adds an element to the property value.
+     *
+     * <p>The given provider will be queried when the value of this property is queried.
+     * This property will have no value when the given provider has no value.
```

Comment:
:question: what would be the way forward if we decide that we want to simply ignore the empty providers? This topic came up several times already and many users even inside the team were surprised by this behavior. I don't think "baby steps" approach to change one aspect of the semantics at a time works here, because we can quickly exhaust the pool of good names. It could work with configurers, of course.

Reply:
>what would be the way forward if we decide that we want to simply ignore the empty providers? 

FTR: we've discussed it extensively and agreed upon "add-atop, ignore-absent" semantic by default, as it seems to be the most common use case. We need to communicate that and see if there are users that prefer existing "add-drops-convention-items, absent-poisons" semantics or other variations of it. We'll gather that feedback over the incubation period.